lib/charms/ip_router_interface/v0/ip_router_interface.py

```diff
@@ -275,208 +263,241 @@ class RouterProvides(Object):
             str([r.name for r in ip_router_relations]),
         )
 
-    def get_all_networks(self) -> List[Network]:
+    def get_routing_table(self) -> RoutingTable:
         """Fetches combined routing tables made available by ip-router providers
 
         Returns:
-            A list of objects of type `Network`. This list contains networks
-            from all ip-router providers that are integrated with the charm.
+            An object of type `RoutingTable` as defined in this file.
         """
         if not self.charm.unit.is_leader():
             return
 
-        router_relations = self.model.relations.get(self.relationship_name)
-        all_networks = []
+        router_relations = self.model.relations.get(self.relation_name)
+        validated_routing_table: RoutingTable = {}
         for relation in router_relations:
-            if networks := relation.data[relation.app].get("networks"):
-                for network in json.loads(networks):
+            if relation_data := relation.data[relation.app].get("networks"):
+                try:
+                    routing_table_from_databag: RoutingTable = json.loads(relation_data)
+                except json.decoder.JSONDecodeError:
+                    logger.error(
+                        "The router's databag has been misconfigured. Can't build routing table."
+                    )
+                    return {}
+
+                for network_name, network_entry in routing_table_from_databag.items():
                     try:
-                        _validate_network(network, {"existing-networks": all_networks})
+                        _validate_network(network_entry, validated_routing_table)
                     except (ValueError, KeyError) as e:
                         logger.warning(
                             "Malformed network detected in the databag:\nNetwork: (%s)\nError: (%s)",
-                            network,
+                            network_entry,
                             e.args[0],
                         )
                     else:
-                        all_networks.append(network)
-            logger.debug(
-                f"Read networks from app: (%s) and relation: (%s)",
-                relation.app.name,
-                self.relationship_name,
-            )
-        return all_networks
+                        validated_routing_table[network_name] = network_entry
+                logger.debug(
+                    "Read networks from app: (%s) and relation: (%s)",
+                    relation.app.name,
+                    self.relation_name,
+                )
+        return validated_routing_table
+
+    def get_network(self, network_name: str) -> Network:
+        """Fetches the network configuration of a specific network.
+
+        Args:
+            network_name:
+                The requested network name
+        Returns:
+            An object of type Network"""
+        return self.get_routing_table()[network_name]
```

Comment:
also there are not tests for this method

Reply:
I've added that info to the docstring. Also, there's no need to test since get_routing_table is tested. A test on this method is equivalent to testing the square bracket operator on a primitive type.